feat: support Pytorch flexible primitives [DET-3194, DET-3195, DET-3196, DET-3197, DET-3198]

shiyuann opened the pull request:
## Description

This PR supports flexible primitives in Pytorch, i.e. supporting multiple optimizers and LR schedulers. It also modifies the current interface to be more Pytorch-y, making porting Pytorch models easier than before.

### Interface

1. Users need to initialize the model, optimizers, and LR schedulers in the constructor of the trial class themselves as opposed to providing a callback for Determined to call in the old interface. Note that all the functions provided in the trial interface start with an underscore. These underscores will be removed when officially releasing this feature. They need to call the wrapping functions of the context object to wrap the model, optimizers, and LR schedulers.

2. Users need to run backward passes and perform one optimizer step for all the optimizers in the train_batch. They can decide in which order to run them.

3. All the interfaces have backward compatibility with the old interface.

Here is a code example:
```
class ModelA(nn.Module):
   def __init__(self, latent_dim, img_shape):
      ...
   def forward(self, z):
      ...

class ModelB(nn.Module):
   def __init__(self, img_shape):
      ...
   def forward(self, img):
      ...

class Model(nn.Module):
   def __init__(self, context):
        self.context = context
        # Initialize 
        mnist_shape = (1, 28, 28)
        self.model_a = ModelA(latent_dim=self.get_hparam("latent_dim"), img_shape=mnist_shape)
        self.model_b = ModelB(img_shape=mnist_shape)

   def forward(self, z):
       return self.model_a(z)

class MultiModelExample(PyTorchTrial):
    def __init__(self, trial_context: det.TrialContext):
        self.context = trial_context

        # Initialize a model with multiple sub-models.
        mnist_shape = (1, 28, 28)
        self.model = self.context._Model(Model(self.context))
        self.model_a, self.model_b = self.model.model_a, self.model.model_b

        # Initialize multiple optimizers with AMP.
        b1, b2 = self.context.get_hparam("b1"), self.context.get_hparam("b2")
        lr = self.context.get_hparam("lr")
        self.opt_a = self.context._Optimizer(torch.optim.Adam(self.model_a.parameters(), lr=lr, betas=(b1, b2)))
        self.opt_b = self.context._Optimizer(torch.optim.Adam(self.model_b.parameters(), lr=lr, betas=(b1, b2)))
        (self.model_a, self.model_b), (self.opt_a, self.opt_b) = self.context.amp_init(
            model=self.model,
            optimizers=[self.opt_a, self.opt_b],
            opt_level="O0",
        )

        # Initialize LR scheduler.
        lambda1 = lambda epoch: epoch // 30
        lambda2 = lambda epoch: 0.95 ** epoch
        self.scheduler1 = self.context._LRScheduler(LambdaLR(self.opt_a, lr_lambda=[lambda1, lambda2]))

    def train_batch(self, batch, epoch_idx: int, batch_idx: int):
        # forward pass for model A
        loss_a=...

        # backward pass for model A
        self.context.backward(loss_a)
        self.context.step_optimizer(self.opt_a)
        self.context.step_lr_scheduler(self.scheduler1)

        # forward pass for model B
        loss_b = ...

        # backward pass for model B
        self.context.backward(loss_b)
        self.context.step_optimizer(opt_b)

        train_metrics = OrderedDict({
            'loss': d_loss,
            'progress_bar': tqdm_dict,
            'log': tqdm_dict
        })

        return train_metrics
```

## Test Plan

The backward compatibility is tested using the test suite in the CI system. Additional tests will be added in a follow-up PR. 

## Commentary (optional)

Please review the second commit. The first commit is in https://github.com/determined-ai/determined/pull/715.
<!-- User-facing API changes need the "User-facing API Change" label -->

<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")

## Description

Lead with the intended commit body in this description field. For breaking changes, please include "BREAKING CHANGE:" at the beginning of your commit body. At minimum, this section should include a bracketed reference to the Jira ticket, e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.

## Test Plan

Describe the situations in which you've tested your change, and/or a screenshot as appropriate. Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.

## Commentary (optional)

Use this section of your description to add context to the PR. Could be for particularly tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc. You may intentionally leave this section blank and remove the title.
--->


[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        check.is_not_none(self.model, "Must have one model")
+        check.gt_eq(len(self.optimizers), 1, "Must have at least one optimizer")
+
+    def _Model(self, model: torch.nn.Module) -> torch.nn.Module:
```
question: why capital letter naming convention?

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            if lr_scheduler is not None:
+                self._LRScheduler(lr_scheduler.scheduler, lr_scheduler.step_mode)
+
+            amp_setting = (
```
blocking: use `self._use_amp()`

shiyuann replied:
This is for backward compatibility of configuring AMP in the experiment configuration. `self._use_amp()` is used after configuring AMP.

aaron276h replied:
Should use `ExperimentConfig.mixed_precision_enabled()` instead here then. Also should avoid ordering requirements between functions when it's not necessary to have them.

aaron276h replied:
> Since we are moving from configuring mixed precision in the experiment config to configure_amp, we will remove mixed_precision_enabled function at the same time. So it is not a good idea to rely on the code of the old interface.

Right, but you would be using it here only to check whether it's configured via the old interface. So once that is deprecated you won't need this check anymore.

shiyuann replied:
DOne.

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            return self.lr_schedulers[0]
+        return None
+
+    def _load_trial(self, trial: "pytorch.PyTorchTrial") -> None:
```
question: once we deprecate the old interface this function is going to go away?

shiyuann replied:
yes

shiyuann replied:
And moved this function to `PytorchTrialController`. Maybe it would be clearer

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        # Initialize LR schedulers. This block must happen after wrapping the optimizers using
+        # horovod.DistributedOptimizer because torch.optim._LRScheduler directly access the
+        # optimizer parameter groups, which are also Horovod.DistributedOptimizer relies on.
+        wrapped = pytorch.LRScheduler(lr_scheduler, step_mode)
```
blocking: we should perform a check there that the optimizer that the scheduler is pointing at is one of the optimizers we returned to the user.

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        optimizers: Union[torch.optim.Optimizer, List[torch.optim.Optimizer]],  # type: ignore
+        mixed_precision: str = "O0",
+    ) -> Tuple[torch.nn.Module, List[torch.optim.Optimizer]]:  # type: ignore
+        self.mixed_precision = mixed_precision
```
blocking: checking mixed precision setting should all be done in `_use_amp()` otherwise you get ordering requirements between the functions.

aaron276h replied:
On second thought, since users will be calling `_configure_amp()` themselves, we should get rid of `_use_amp()` function, and instead have a private bool `self._using_amp` that gets set to `True` if `_configure_amp()` is called. We also should get rid of the  `if mixed_precision != "O0":` check below, if a user wants to configure amp with "O0" we should let them. However, until we switch over to the new API, we should call `_configure_amp()` ourselves only when `ExperimentConfig.mixed_precision_enabled()` is `True`.

shiyuann replied:
So if we allow setting `O0` in mixed precision not take it as mixed precision unset. We need to change the master's logic of setting mixed_precision as default. This involves a large change in the whole cluster. I'd leave to #ml-ag to figure this out.

shiyuann replied:
Done. But I wouldn't use `ExperimentConfig.mixed_precision_enabled()` since it will be removed along with supporting AMP in the constructor.

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        for p in filter(lambda param: param.grad is not None, parameters):
+            p.grad.data.div_(divisor_value)
+
+    def _step_optimizer(self, optimizer: torch.optim.Optimizer) -> None:  # type: ignore
```
question: we had previously talked about handing the optimizer a closure to execute for gradient clipping, is that outside the scope of this PR?

shiyuann replied:
Sorry forgot that. What do you mean by `closure`, passing in a function?

shiyuann replied:
Added an argument to pass in a function that accepts the parameters as an argument and backward compatibility for `on_before_optimizer_step` callback.

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        return model, optimizers
+
+    def _should_communicate_and_update(self) -> bool:
+        if self.current_batch_idx is None:
```
non-blocking: could we make getters and setters for `self.current_batch_idx`

shiyuann replied:
We don't provide functions to set `current_batch_idx` to users and it's used only in one place. Sound unnecessary to have getters and setters

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            # Communication needs to be synchronized so that is completed
+            # before we apply gradient clipping and `step()`.
+            if self.hvd_config.use and self._should_communicate_and_update():
+                for optimizer in self.optimizers:
```
non-blocking: i would consider removing all support for multiple optimizers until we make it fully supported.

shiyuann replied:
What do you mean it's not fully supported?

aaron276h replied:
E.g., our train_batch loop and saving/restoring logic does not support multiple optimizers

shiyuann replied:
This will be supported in this PR. I updated the name of this PR.

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        self,
+        model: torch.nn.Module,
+        optimizers: Union[torch.optim.Optimizer, List[torch.optim.Optimizer]],  # type: ignore
+        mixed_precision: str = "O0",
```
blocking: we should be exposing all the args of `amp.init()` to users as args of this function.

shiyuann replied:
Done

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+
+    def _configure_amp(
+        self,
+        model: torch.nn.Module,
```
non-blocking: We should consider supporting the same options as amp.initialize here.  @aaron276h already mentioned exposing a bunch of options, but there two other important differences:
  - apex.amp.initialize() allows you to pass in a single model or a list of models
  - apex.amp.initialize() allows you to pass in None for the optimizer

https://github.com/NVIDIA/apex/blob/master/apex/amp/frontend.py#L195

I marked this as non-blocking because I don't think it needs to block *this* PR, but I would block on this before exposing this to the API public.

shiyuann replied:
Since we are supporting multiple models via sub-models of the main model, I think we should only let users pass in only one model. But we should let users pass in at least one optimizer since `apex` supports None only for inference.

shiyuann replied:
I have added the arguments of `amp.initialize` to `_configure_amp`. So will resolve this conv.

rb-determined-ai replied:
(Unresolving this, please do not resolve a thread before I have a chance to see your response)

> Since we are supporting multiple models via sub-models of the main model

Is this true just for right now, or is it true in the state of the final API as well?  I thought our planned API worked just fine with multiple separate models.

shiyuann replied:
It's true for the final API as well. Implementing multiple sub-models would be very simple for users and implementing multiple models would be hacky for us. Pytorch `nn.Module` sees a member field `self.sub_model=model_a` as a first-class citizen compared to `self.sub_model[0]=model_a`. The latter can't be recognized automatically by the `nn.module` implementation for serialization, deserialization, parameters, buffers, and etc.

shiyuann replied:
We have had a discussion for multiple sub-models API and it seems fine for internal users. https://determined-ai.slack.com/archives/C0147BBR61E/p1591306980029400

shiyuann replied:
Discussed offline. Will go with multiple models.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+    def _use_amp(self) -> bool:
+        return self.mixed_precision != "O0"
+
+    def _configure_amp(
```
question: What do you think about having the name `apex_amp` in this function name?  I'm concerned that if we don't put `apex_amp` in the name, then people will be confused in just a few months when pytorch has their own `amp` as well.

shiyuann replied:
I think generally we should be opinionated on what library to use internally rather than giving two choices to users and exposing configuration details to users.

rb-determined-ai replied:
Ok, let's consider the possibilities:

If we name this `configure_amp()` (or something non-apex-specific):
- we are allowed to only support apex amp if we think that is good. (good)
- future users might not realize that we are talking about apex amp instead of the torch-native amp. (bad)
- we will have naming problems if we later decide to support apex amp AND torch-native amp.  We can basically guarantee that we will have users who ask for both at some point. (bad)

If we name this `configure_apex_amp()` (or something that is apex-specific):
- we are allowed to only support apex amp if we think that is good. (good)
- future users will know we are talking about apex amp instead of the torch-native amp. (good)
- we will not have any naming problems if we later decide to support apex amp AND torch-native amp. (good)

To me, choosing to name this something non-apex-specific when we know that there's going to be two amp options in the ecosystem very soon (pytorch 1.6) seems like we are opening ourselves up to future problems.

shiyuann replied:
okay. @aaron276h What are your thoughts on this name?

aaron276h replied:
I would go with `configure_apex_amp()`, that way if we decide to support both methods in the future, the API will make it clean to do so.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            logging.debug("Initialized optimizer for distributed and optimized parallel training.")
+
+        self.optimizers.append(optimizer)
+        return optimizer
```
We should consider modifying their optimizer to fail if they call `optimizer.zero_grad()` themselves, or ignore it but issue a warning or something.

shiyuann replied:
Yes. I thought so, too. However, we should create a new class for optimizers as opposed to monkey patching the optimizers.

rb-determined-ai replied:
Horovod already creates a class dynamically around the original optimizer, so I am not confident that it makes a lot of difference here.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+                    parameters=parameters, divisor=self.hvd_config.aggregation_frequency
+                )
+
+            for callback in self.callbacks.values():
+                callback.on_before_optimizer_step(parameters)
```
question: are we going to keep this callback in the final API?

shiyuann replied:
Want to remove this in the final API. Added a comment for future removing.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -100,55 +79,12 @@ def from_trial(*args: Any, **kwargs: Any) -> det.TrialController:
     @staticmethod
     def from_native(*args: Any, **kwargs: Any) -> det.TrialController:
-        raise NotImplementedError("PyTorchTrial only supports the Native API")
+        raise NotImplementedError("PyTorchTrial only supports the Trial API")
```
lol

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -749,23 +589,43 @@ class PyTorchTrial(det.Trial):
     trial_context_class = PyTorchTrialContext
 
-    @abstractmethod
+    def __init__(self, trial_context: PyTorchTrialContext) -> None:
```
blocking: we should not have a default `__init__()`.  Users *must* define their own.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -776,6 +636,29 @@ def train_batch(
         """
         pass
 
+    def _train_batch(
```
blocking: we should not be putting any logic as part of the PytorchTrial class, this should live as part of our own pytorch framework logic on TrialController.

rb-determined-ai commented on harness/tests/experiment/utils.py:
```diff
@@ -413,7 +413,13 @@ def create_trial_instance(trial_def: Type[det.Trial]) -> None:
         trial_instance = experimental.create_trial_instance(
             trial_def=trial_def,
-            config={"hyperparameters": {"global_batch_size": det.Constant(16)}},
+            config={
```
What is this change for?

shiyuann replied:
Unfortunately, we break this test by creating model and optimizers in the constructor, which requires these hparams. We used not to create them in the test.

shiyuann commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            raise det.errors.InternalException("Training hasn't started.")
+        return (self.current_batch_idx + 1) % self.hvd_config.aggregation_frequency == 0
+
+    def _backward(self, loss: torch.Tensor):  # type: ignore
```
Done.

shiyuann commented:
@rb-determined-ai @aaron276h I updated the code changes according to your comments. I also corrected the use of `amp.scale_loss` for multiple losses, which requires to specify the number of losses in `amp.initialize` and a loss id for each `amp.scale_loss`.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -17,11 +32,23 @@ class PyTorchTrialContext(det.TrialContext):
+        self.callbacks = {}  # type:  Dict[str, pytorch.PyTorchCallback]
+
+        self.scaler_mapping = {}  # type: Dict[torch.Tensor, int]
+        self.loss_num = 0
```
blocking: rename to something a bit clearer, like `next_loss_id` or `next_apex_loss_id`

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +85,311 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+
+            # Communication needs to be synchronized so that is completed
+            # before we apply gradient clipping and `step()`.
+            if self.hvd_config.use and self._should_communicate_and_update():
```
question: shouldn't this get called as part of `_step_optimizer()`?  It seems like this is going to behave incorrectly in the case that the user calls `_backward()` on more than one loss.

shiyuann replied:
Right. Nice catch..

aaron276h replied:
Yeah this is going to perform communication for every call to `backwards`: https://github.com/horovod/horovod/blob/9de29ca469a57578b1098d20ac6f4262315be21a/horovod/torch/optimizer.py#L137

The tricky case is with amp, since we need to perform the synchronization before unscaling which happens at `__exit__`.

shiyuann replied:
Discussed offline:

Because the library apex we are relying on to support AMP isn’t 100% compatible with horovod, so the implementation of this project might slow down the training if users use AMP, d-train, multiple losses, and multiple optimizers at the same time. The degree of the slowdown depends on the number and sizes of parameters to synchronize. In the future, we will support Torch native AMP, which is not in the scope of this project. We will wait until then to resolve this issue.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -20,55 +19,97 @@
+                self.context._step_optimizer(self.context.optimizers[0], clip_grads=clip_grads)
+                return tr_metrics
+
+            self.trial.__setattr__("train_batch", new_train_batch)
```
praise: I really, really like this strategy for being backwards compatible.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -20,55 +19,97 @@
+            train_batch = self.trial.train_batch
+
+            def new_train_batch(
+                batch: TorchData, model: nn.Module, epoch_idx: int, batch_idx: int
```
question: at some point we are going to take the `model` argument out, right?  When is the right time to do that?

shiyuann replied:
Since we only have one main model for each trial class, I think it might make sense to keep this argument. Don't see any additional benefits to make changes so I'd rather not make any changes

shiyuann replied:
Update: will remove model in the future since we are introducing multiple models

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -679,16 +562,33 @@ def _load(self) -> None:
-        if self.context.lr_scheduler is not None:
-            self.context.lr_scheduler.load_state_dict(checkpoint.get("lr_scheduler"))
+
+        # TODO(DET-3262): remove this backward compatibility of supporting the old schema
+        #     of checkpointing the only optimizer.
```
We should plan on supporting loading old models for... well for quite a while.  We have customers with old checkpoints in production, and they would not appreciate if we broke that for them.  The precise answer of "how long" is probably worth a larger discussion.

shiyuann replied:
Yes. I don't know the answer. In this case, I guess my answer for this question is unless we have to change we don't change. I guess I'll remove this comment because we shouldn't do that in `DET-3262`

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -17,11 +32,23 @@ class PyTorchTrialContext(det.TrialContext):
+
+        self._use_amp = False
+
+        self.current_batch_idx = None  # type: Optional[int]
```
non-blocking: make this private unless it's being used outside the class.

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +85,311 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        return model
+
+    def _Optimizer(self, optimizer: torch.optim.Optimizer) -> torch.optim.Optimizer:  # type: ignore
+        check.is_not_none(self.model, "Must have one model first.")
```
non-blocking: this message makes it sound like users may specify more than model.

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +85,311 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        return None
+
+    def _Model(self, model: torch.nn.Module) -> torch.nn.Module:
+        check.is_none(self.model, "Must call Model() only once and before calling configure_amp.")
```
blocking: I would expand this error message explaining to users how to they use multiple models via subclassing or point to docs at least. I would drop the configure_amp part of this message.

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +85,311 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        """
+        step_optimizer performs a single optimization step.
+
+        This function must be called once for each optimizer. However, the order of
```
blocking: can you add Args docs for this

shiyuann commented:
@aaron276h It's outside the scope of this PR.

The test plan is:

1. Update all the unit tests in `test_pytorch_trial.py` for multiple optimizers and LR schedulers, including tests for training metrics, checkpointing, restoring, reproducibility, and gradient clipping.
2. Update the e2e tests for parallel training with warm starting, different AMP settings, and aggregation frequency.
3. Add unit tests for wrong initialization

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +84,356 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        # TODO: Check that optimizer is not an amp optimizer.
+        if self.hvd_config.use:
+
+            def filter_named_parameters(
+                model: nn.Module, optimizer: torch.optim.Optimizer,  # type: ignore
+            ) -> List:
+                """
+                filter_named_parameters filters the named parameters of a specified optimizer out
+                of all the named parameters from a specified model. We need this function because
+                a torch.optim.Optimizer doesn't store parameter names and we need the names of
+                the parameters when mapping parameters to each horovod.DistributedOptimizer.
+                """
+                opt_params = {
+                    p for group in optimizer.param_groups for p in group.get("params", [])
+                }
+                return [(name, p) for name, p in model.named_parameters() if p in opt_params]
+
+            use_compression = self.hvd_config.fp16_compression
+            optimizer = hvd.DistributedOptimizer(
+                optimizer,
+                named_parameters=filter_named_parameters(cast(nn.Module, self.model), optimizer),
+                backward_passes_per_step=self.hvd_config.aggregation_frequency,
+                compression=hvd.Compression.fp16 if use_compression else hvd.Compression.none,
+            )
+            logging.debug("Initialized optimizer for distributed and optimized parallel training.")
```
This is unpleasant enough that I think we should weight the pros and cons of the API against what we have to do behind the scenes.

So the reason for this, as @shiyuann explained to me, is that horovod needs the named parameters.  Horovod expects the user to feed the `named_parameters` into the horovod optimizer, and presumably they need the to match the `parameters` that were fed to the optimizer:

```python
optimizer = optim.SGD(model.parameters())
optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters())
```

The fact is that you can go from named_parameters to parameters, but the reverse lookup is not natively supported.

So I see three options:

1. do a reverse lookup of the named_parameters from the parameters based on our list of all the models (the current PR implementation)
2. mimic the horovod API and make users hand us the named_parameters explicitly
3. we hook into the optimizer initialization and users hand us named_parameters

### 1. do a reverse lookup

(code example is just basically the current implementation)

pros:

- API is super easy to use

cons:

- It's not clear to me if there are use cases that we are going to be unable to support.  I imagine that if the parameters in an optimizer were not already part of one of the registered models, something else would probably break anyway, so this might not be so bad.

### 2. minimc the horovod API

```python
optimizer = optim.SGD(model.parameters())
optimizer = context._Optimizer(optimizer, named_parameters=model.named_parameters())
```

pros:

- Solution is pretty bulletproof; if horovod supports the training then we would too, because we're essentially passing the horovod API straight to the user.

cons:

- The API is harder to use, and users who don't care about dtrain still need to adjust their code to fit a dtrain paradigm

### 3. hook into the optimizer initialization

```python
# under the hood, we instantiate the optimize by stripping names from the named_parameters
optimizer = context._make_optimizer(optim.SGD, model.named_parameters())
```

pros:

- Literally every `Optimizer` in `torch` takes parameters as its first `__init__` parameter, so this API would be reasonably natural.
- No reverse lookup necessary

cons:

- This would kind of create a weird API that I wouldn't love if I was a user.  I think wrapping constructed objects is more natural than wrapping constructed.

### Conclusion

I actually changed my own mind while writing this.  I started thinking that option 3 was the best but now I think that option 1 (what you already have) is the best, unless anyone (@aaron276h, @katport) can think of a case where the parameters of an optimizer are not part of one of the models (since that would break the reverse lookup).

@shiyuann  still, why don't we simplify the reverse lookup by building a dictionary from all of the models we have registered in the context (see below)?  It seems like this would be a very clean mechanism and wouldn't require users to build all of their models as a single model with submodels.

```python
class PyTorchContext(TrainContext):
     def __init__(self, ...):
          self._parameter_names = {}

     def _Model(self, model):
          ...
          for name, param in model.named_parameters():
              self._parameter_names[parameter] = name
     
     def _Optimizer(self, opt):
          ...
          named_parameters = {self._named_parameters[p] for p in opt_params}
          optimzer = hvd.DistributedOptimizer(
               opt,
               named_parameters = named_parameters,
               ...
           )
           ...
```

shiyuann replied:
>  @shiyuann still, why don't we simplify the reverse lookup by building a dictionary from all of the models we have registered in the context (see below)? It seems like this would be a very clean mechanism and wouldn't require users to build all of their models as a single model with submodels.

Because `named_parameters()` is way more complex than simply just adding parameters to a dict. You need to consider the cases if there are sub-modules refer to the same module. Basically, we'll have to duplicate the logic in `nn.Module.named_modules()`.

aaron276h replied:
If we decide to go down the path of supporting multiple models (as I understand the current agreed on plan required users to submit one model that they have multiple modules inside of) I think the most transparent approach is requiring named_parameters (approach #2).

rb-determined-ai replied:
@shiyuann, I think we need to understand what Horovod needs the named_parameters for in the first place.

My gut sense was pytorch optimizers keep a list of parameters for tracking gradient updates (no name needed since they can just keep the objects in memory), and the only reason that horovod needs named parameters is to know which parameters in one process' memory space correspond with parameters in the other proccess' memory space.  This turns out to be true (https://github.com/horovod/horovod/blob/master/horovod/torch/optimizer.py#L32)

Horovod already handles the case where multiple names point to a single parameter; they just sort the list of named parameters by name when they build their reverse-lookup dictionary (which they actually call `self.parameter_names` too haha): https://github.com/horovod/horovod/blob/master/horovod/torch/optimizer.py#L64.  Admittedly, the solution I proposed isn't quite right, but it's not very far from right either.

@aaron276h I agree that option 2 is more transparent, but is there any possible situation where a user would have a trainable tensor that is not part of one of the models that they register with us, but where we still support their use case?  I think we can safely hide this named_parameter details from users since we already require users to register every model in their system with us.  This is different than native Horovod, since native Horovod only actually hooks into the optimizer.

@shiyuann, @aaron276h There's really two questions here:

A. Is there a technical reason why we need to enforce the single-model-with-submodels strategy on users?  (this was the original motivation for this investigation)
B. What should our `context._Optimizer()` api look like?

Regarding A, I don't see any technical reason why we need to enforce the single-model-with-submodels strategy on users.  I think that would be a pretty arbitrary requirement of our API.

Regarding B, @aaron276h, does what I said here change your mind?

shiyuann replied:
< 
@aaron276h I agree that option 2 is more transparent, but is there any possible situation where a user would have a trainable tensor that is not part of one of the models that they register with us, but where we still support their use case? I think we can safely hide this named_parameter details from users since we already require users to register every model in their system with us. This is different than native Horovod, since native Horovod only actually hooks into the optimizer.

I think we can safely assume all the parameters that need to update are registered in at least one module.

shiyuann replied:
For question A, as I said in other threads, it's not blocking but it makes the implementation unnecessarily complex. 

It will be:

```
params_a = model_a.named_parameters()
params_b = model_a.named_parameters()
all_params = {}
# remove duplicate module
params = set()
prefix="params_a"
for k, v in params_a:
    if v in params:
         pass
    params.add(v)
    params[prefix+k] = v

# Do the same for all the modules.
...
# Find the parameters of the optimizer in all the parameters.
...
```

shiyuann replied:
Update after discussing it offline: we will go with multiple models

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -750,19 +661,62 @@ class PyTorchTrial(det.Trial):
 
     @abstractmethod
+    def __init__(self, trial_context: PyTorchTrialContext) -> None:
+        """
```
blocking: we can't add this to the API documentation before we have officially released the API.  As I understand it, this API is not going to be officially released until some time after this PR lands, so this docstring should land at that time.

I *think* that the docstrings for private methods like `_Model` don't get written automatically, so those doc strings would be fine.  But you should double check that.

shiyuann replied:
Will move this docstring to `PytorchTrialContext`

shiyuann replied:
> blocking: we can't add this to the API documentation before we have officially released the API. As I understand it, this API is not going to be officially released until some time after this PR lands, so this docstring should land at that time.

I think it's fine to add documentation here and call it internal/soft release since all the functionalities are in this PR. I think the underscores can be removed when adding documentation for the new interface since it would be weird to let users use functions starting with underscores in the doc. On the official release, which happens after two rounds of dogfooding, we will deprecate the use of the old interface. We don't need to wait until that time.

shiyuann replied:
I think the release plan should be updated per the new API because the old interface proposal is not as invasive as the new one. Here are some thoughts on the constraint and needs we have.

1. Any update on the docstring will show up on the documentation.
2. I don't want to update the docstring until releasing it otherwise the docstring would be inconsistent with the implementation.
3. It's fine to release early before we could tell users we prefer this method. It could say in the documentation. it's a different approach from the current one and it's still experimental.
4. It doesn't make sense to put the interface in `experimental` namespace since it's an improvement on the current interface as opposed to being a risky feature and it's preferred over the older interface. An example that uses a different tactic is the Native project, which is risky and has a large space to improve before we are able to let it replace Trial API.
5. We want to deprecate the old method after internal dogfooding.

The previous plan is to put underscores in the interface functions before releasing it because the training code update might be in different PRs and the new functions would be actually useless.

So I think it's fine to put documentation in the docstring and remove the underscores in this PR since this PR contains the implementation of training for the new interface. I'll remove the underscores and add more docs to the `PytorchTrial` and `PytorchTrialContext` in this PR.

rb-determined-ai replied:
To summarize the conversation we had offline, I really don't think it's a good idea to publicly expose an API while we *fully expect* to rename some or possibly all of the functions.

shiyuann commented on common/determined_common/experimental/checkpoint/_torch.py:
```diff
@@ -17,6 +17,6 @@ def load_model(ckpt_dir: pathlib.Path, metadata: Dict[str, Any], **kwargs: Any)
     model = trial.build_model()
     checkpoint = torch.load(ckpt_dir.joinpath("state_dict.pth"), map_location="cpu")  # type: ignore
-    model.load_state_dict(checkpoint["model_state_dict"])
+    model.load_state_dict(checkpoint["models_state_dict"][0])
```
No we haven't but I'll create a ticket for that.

shiyuann commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +85,335 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        Optimizer creates a horovod.DistributedOptimizer if uses parallel/distributed training.
+        """
+
+        # TODO: Check that optimizer is not an amp optimizer.
```
I don't know how to check if an optimizer is an amp optimizer. But we can check `self._use_amp` in `Model` and `Optimzier` to make sure `configrue_apex_amp` happen after them.

rb-determined-ai replied:
that would be fine I think.

shiyuann commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +85,335 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            logging.debug("Initialized mode for native parallel training.")
+
+        model_id = len(self.models)
+        self._main_model.__setattr__(f"model_{model_id}", model)
```
I have tried `named_parameters` with a prefix, which unfortunately fail. The place where it fails is when we call `hvd.broadcast_parameters(model.state_dict())`. In the docstring of `broadcast_parameters`, it says you could use `model.named_parameters()` but I got weird errors so I switched back to use `state_dict`. `state_dict` is calculated with parameters and buffers and some internal state. So I choose to use this way.

rb-determined-ai replied:
sgtm

aaron276h replied:
non-blocking: can you add a comment about why this is necessary

shiyuann commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +85,335 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        self,
+        models: Union[torch.nn.Module, List[torch.nn.Module]],
+        optimizers: Union[torch.optim.Optimizer, List[torch.optim.Optimizer]],  # type: ignore
+        enabled: Optional[bool] = True,
+        opt_level: Optional[str] = "O1",
+        cast_model_type: Optional[torch.dtype] = None,
+        patch_torch_functions: Optional[bool] = None,
+        keep_batchnorm_fp32: Optional[Union[bool, str]] = None,
+        master_weights: Optional[bool] = None,
+        loss_scale: Optional[Union[float, str]] = None,
+        cast_model_outputs: Optional[torch.dtype] = None,
+        num_losses: Optional[int] = 1,
+        verbosity: Optional[int] = 1,
+        min_loss_scale: Optional[float] = None,
+        max_loss_scale: Optional[float] = 2.0 ** 24,
```
I think we simply call `apex.amp.init` for the users. Don't know if there is any arguments that are not supported. We got feedbacks on the implementation of Keras `WrappedModel.fit` arguments that the use of `args` and `kwargs` prevent users from using autofill. I think let's duplicate all the arguments here and see if it actually is helpful.

rb-determined-ai replied:
sgtm

shiyuann commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +85,335 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+                products. Defaults to ``False``.
+        """
+        if self._use_amp:
+            if loss not in self._scaler_mapping:
```
Will fix it by adding checks.

shiyuann commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -17,11 +32,23 @@ class PyTorchTrialContext(det.TrialContext):
+        self.models = []  # type: List[nn.Module]
+        self.optimizers = []  # type: List[torch.optim.Optimizer] #  type: ignore
+        self.lr_schedulers = []  # type: List[pytorch.LRScheduler]
+        self.callbacks = {}  # type:  Dict[str, pytorch.PyTorchCallback]
```
okay. You mean `TrialController`, right?

rb-determined-ai replied:
yes, mibad

shiyuann commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -356,93 +295,43 @@ def _train_for_step(self, step_id: int, batches_per_step: int) -> workload.Respo
                 dict,
                 "train_batch() must return a dictionary "
-                "mapping string names to Tensor metrics, got {type(tr_metrics)}",
+                f"mapping string names to Tensor metrics, got {type(tr_metrics)}",
```
yeah. and you praised me twice for this. lol

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        return None
+
+    def Model(self, model: torch.nn.Module) -> torch.nn.Module:
+        """Initialize a model. It returns an initialized model."""
```
As before, we should avoid calling this "initialize" a model, since the model is actually already initialized.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        return model
+
+    def Optimizer(self, optimizer: torch.optim.Optimizer) -> torch.optim.Optimizer:  # type: ignore
+        """Initialize an optimizer. It returns an initialized optimizer.
```
ditto

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        lr_scheduler: torch.optim.lr_scheduler._LRScheduler,
+        step_mode: pytorch.LRScheduler.StepMode,
+    ) -> pytorch.LRScheduler:
+        """Initialize a LR scheduler. It returns an initialized LR scheduler.
```
ditto

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            self.device = torch.device("cpu")
+        check.is_not_none(self.device)
+
+    def to_device(self, data: pytorch._Data) -> pytorch.TorchData:
```
Why are we making this public?

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        max_loss_scale: Optional[float] = 2.0 ** 24,
+    ) -> Tuple:
+        """
+        Configure automatic mixed precision for your model and optimizers. Note that Determined
+        integrates apex.amp internally for AMP features.
```
What do you think about replacing:

"Note that Determined integrates apex.amp internally for AMP features"

with:

"Note that details for apex.amp are handled automatically within Determined after this call"

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        Configure automatic mixed precision for your model and optimizers. Note that Determined
+        integrates apex.amp internally for AMP features.
+
+        This function should be called **after** you have finished constructing your model and
+        optimizer(s) with :meth:`Model` and :meth:`Optimizer`.
```
This function ~should~ must ...

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        check.eq(
+            num_losses,
+            1,
+            "Only support configure apex amp on distributed training with only one loss",
```
Is this restriction only for distributed training?  I thought we had this restriction in all cases?

In any case, how about rephrasing to:

"Determined only supports apex.amp with num_losses = 1"

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        )
+        check.false(
+            not self.hvd_config.use and self.n_gpus > 1,
+            "To enable mixed precision training (AMP) for parallel training, "
+            'please set `resources["optimized_parallel"] = True`.',
```
Wait, I thought we removed `optimized_parallel` entirely?

aaron276h replied:
Yeah this error message is very out of date

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            check.eq(
+                loss,
+                self._loss,
+                "Only support configure apex amp on distributed training with only one loss",
```
"Calling context.backward() multiple times is not supported while using apex.amp"

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        optimization.aggregation_frequency.
+
+        Arguments:
+            optimizer(``torch.optim.Optimizer``): An optimization step of which optimizer should
+                be performed.
```
"An optimization step of which optimizer should be performed" -> "which optimizer should be stepped"

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -20,57 +19,102 @@
         if self.hvd_config.use:
-            hvd.broadcast_parameters(self.context.model.state_dict(), root_rank=0)
-            hvd.broadcast_optimizer_state(self.context.optimizer, root_rank=0)
+            logging.info("Broadcasting model parameters")
```
Is this new?  Why do we need to add it?

rb-determined-ai commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -20,57 +19,102 @@
+                hvd.broadcast_optimizer_state(optimizer, root_rank=0)
 
         self.training_iterator = iter(self.training_loader)
+        logging.info("Ready for training")
```
ditto

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        ``torch.optim.Optimizer`` directly access the model parameters,
+        which are initialized in :meth:`Model`.
+
+        Optimizer creates a ``horovod.DistributedOptimizer`` if uses parallel/distributed
```
blocking: uses -> using

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        """
+        if self._use_amp:
+            check.eq(
+                loss,
```
question: won't this check always fail on the first call since `self._loss` is going to be `None`?

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+
+            with apex.amp.scale_loss(loss, self.optimizers) as scaled_loss:
+                scaled_loss.backward(
+                    gradient=gradient, retain_graph=retain_graph, create_graph=create_graph
```
non-blocking: can you add a commment about the `retain_graph` and `create_graph` parameters are needed?

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+                    #
+                    # Unfortunately, there is no clean way to support unscaling
+                    # happening after synchronizing the optimizer on apex.
+                    # A short-term solution is to synchronize all of the optimizers
```
question: this comment needs to be updated right?

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -1,27 +1,72 @@
+       configured with apex amp.
+    4. Calculate the gradients with :meth:`backward` on specific losses.
+    5. Update the parameters with :meth:`step_optimizer`.
+
```
I was apparently in a rush when I wrote this, but by "distributed training data" what I meant was talk about the capabilities of the "DistributedTrainContext".

shiyuann replied:
You mean `DistributedContext`? I mention that in a separate point. But if I remember correctly, this is not a fully supported feature. It's just a short term solution and not even documented.

shiyuann commented:
Updated on the mixed precision support:
1. When using mixed precision and distributed training, we don't support multiple backward passes, including multiple losses and multiple backward passes on a single loss, because of the communication cost.
2. When using distributed training, we don't support multiple backward passes on any parameter, including multiple losses containing a parameter, multiple backward passes on a single loss, because horovod doesn't support manual gradient accumulation.

rb-determined-ai commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +102,359 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+
+            When using distributed training, we don't support manual gradient accumulation.
+            That means the gradient on each parameter can only be calculated once on each batch.
+            If a parameter is associated with multiple losses, you can only call backward on
+            these losses for no more than once. Or you could set require_grads flag of a
```
"If a parameter is associated with multiple losses, you can either choose to call backward on only one of those losses, or you could set the ``require_grads`` flag ..."

"However you could do gradient accumulation across batches by ..."

(the second suggestion is courtesy of @aaron276h)

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +102,359 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+                if self.hvd_config.use and self._should_communicate_and_update():
+                    # When using horovod, we need to finish communicating gradient
+                    # updates before they are unscaled which happens when we exit
+                    # of this context manager.
```
blocking: when we exit out of this ...

shiyuann replied:
Done.

aaron276h commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -57,4 +102,359 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            clip_grads(a function, optional): This function should have one argument for
+                parameters in order to clip the gradients
+        """
+
```
question: is there any mechanism for checking if users have called backwards for all their optimizers once per step? And what happens if they skip an optimizer?

shiyuann replied:
There is no mechanism for that. We don't support skipping optimizer.step. That is equivalent to manual gradient accumulation. Will add checks for that.

shiyuann replied:
Nice catch!

shiyuann replied:
But the horovod will throw the same errors when users skip calling `optimizer.step` as when users do manual gradient accumulation.

aaron276h commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -740,28 +647,95 @@ def _save(self, path: pathlib.Path) -> workload.Response:
     trial_controller_class = PyTorchTrialController
     trial_context_class = PyTorchTrialContext
 
     @abstractmethod
```
question: why is the `_init_` an abstract method?

shiyuann replied:
It's always an abstract method for all the frameworks. For the new interface, we need users to wrap models, optimizers in `__init__` so it's mandatory.
